data/org.freedesktop.impl.portal.Screenshot.xml

```diff
@@ -49,7 +49,9 @@
             <term>interactive b</term>
             <listitem><para>
               Hint whether the dialog should offer customization before taking a screenshot.
-              Defaults to no.
+              Defaults to no. Since version 3 of #org.freedesktop.portal.Screenshot,
+              non-interactive screenshots are mediated by the permission store, so
+              implementations should not show any additional interaction.
```

Comment:
Thinking some more about this, it'd be more correct if `org.freedesktop.impl.portal.Screenshot` had a version set by the backend. That'd avoid breaking backends when the frontend is upgraded.

Reply:
A quick test shows that

```diff
diff --git a/data/org.freedesktop.impl.portal.Screenshot.xml b/data/org.freedesktop.impl.portal.Screenshot.xml
index cd769b5..1ae1608 100644
--- a/data/org.freedesktop.impl.portal.Screenshot.xml
+++ b/data/org.freedesktop.impl.portal.Screenshot.xml
@@ -105,5 +105,6 @@
       <annotation name="org.qtproject.QtDBus.QtTypeName.Out1" value="QVariantMap"/>
       <arg type="a{sv}" name="results" direction="out"/>
     </method>
+    <property name="version" type="u" access="read"/>
   </interface>
 </node>
diff --git a/src/screenshot.c b/src/screenshot.c
index a28339c..ab13a6a 100644
--- a/src/screenshot.c
+++ b/src/screenshot.c
@@ -206,6 +206,9 @@ handle_screenshot_in_thread_func (GTask *task,
   parent_window = ((const char *)g_object_get_data (G_OBJECT (request), "parent-window"));
   options = ((GVariant *)g_object_get_data (G_OBJECT (request), "options"));
 
+  if (xdp_dbus_impl_screenshot_get_version (impl) == 0)
+    goto query_impl;
+
   permission = get_permission_sync (app_id, PERMISSION_TABLE, PERMISSION_ID);
 
   if (!g_variant_lookup (options, "interactive", "b", &interactive))
@@ -286,6 +289,8 @@ handle_screenshot_in_thread_func (GTask *task,
         }
     }
 
+query_impl:
+
   impl_request =
     xdp_dbus_impl_request_proxy_new_sync (g_dbus_proxy_get_connection (G_DBUS_PROXY (impl)),
                                           G_DBUS_PROXY_FLAGS_NONE,
```

here, and

```diff
diff --git a/src/screenshot.c b/src/screenshot.c
index 1dca27d..27fab30 100644
--- a/src/screenshot.c
+++ b/src/screenshot.c
@@ -304,6 +304,7 @@ screenshot_init (GDBusConnection *bus,
   GDBusInterfaceSkeleton *helper;
 
   helper = G_DBUS_INTERFACE_SKELETON (xdp_impl_screenshot_skeleton_new ());
+  xdp_impl_screenshot_set_version (XDP_IMPL_SCREENSHOT (helper), 1);
 
   g_signal_connect (helper, "handle-screenshot", G_CALLBACK (handle_screenshot), NULL);
   g_signal_connect (helper, "handle-pick-color", G_CALLBACK (handle_pick_color), NULL);
```

in xdg-desktop-portal-gnome avoids any regressions.

Reply:
Also, adding a `skip_dialog` boolean instead of the reverse version check might be a bit cleaner too?

Reply:
> Also, adding a skip_dialog boolean instead of the reverse version check might be a bit cleaner too?

I've just quickly discussed this with Allan and my understanding is that this should have been a permission from the get-go; the dialog on the non-interactive case was a mistake. The expectation is that interactive=false will only be set by apps that are themselves screenshot editors, like Flameshot. Consequently, there shouldn't be any case where an app would request non-interactive screenshots when they aren't editors themselves. So I don't think this should be supported from the API.

Reply:
`skip_dialog` would only be part of the impl portal, as a backward compatible way to the new behavior, it wouldn't affect anything app facing.

Reply:
To put my word into diffs, it'd be

```diff
diff --git a/src/screenshot.c b/src/screenshot.c
index ab13a6a..5dc7bc2 100644
--- a/src/screenshot.c
+++ b/src/screenshot.c
@@ -197,6 +197,7 @@ handle_screenshot_in_thread_func (GTask *task,
   gboolean interactive;
   const char *parent_window;
   const char *app_id;
+  gboolean skip_dialog = FALSE;
 
   REQUEST_AUTOLOCK (request);
 
@@ -289,6 +290,8 @@ handle_screenshot_in_thread_func (GTask *task,
         }
     }
 
+  skip_dialog = TRUE;
+
 query_impl:
 
   impl_request =
@@ -310,6 +313,12 @@ query_impl:
                       screenshot_options, G_N_ELEMENTS (screenshot_options),
                       NULL);
 
+  if (skip_dialog)
+    {
+      g_variant_builder_add (&opt_builder, "{sv}", "skip_dialog",
+                             g_variant_new_boolean (TRUE));
+    }
+
   g_debug ("Calling Screenshot with interactive=%d", interactive);
   xdp_dbus_impl_screenshot_call_screenshot (impl,
                                             request->id,
```

and

```diff
diff --git a/src/screenshot.c b/src/screenshot.c
index 27fab30..caed19a 100644
--- a/src/screenshot.c
+++ b/src/screenshot.c
@@ -39,8 +39,6 @@ typedef struct {
   const char *retval;
 } ScreenshotDialogHandle;
 
-static uint32_t screenshot_portal_version = 0;
-
 static void
 screenshot_dialog_handle_free (gpointer data)
 {
@@ -158,6 +156,7 @@ handle_screenshot (XdpImplScreenshot *object,
   ScreenshotDialogHandle *handle;
   gboolean modal;
   gboolean interactive;
+  gboolean skip_dialog;
   GtkWindow *dialog;
   GdkSurface *surface;
   GdkDisplay *display;
@@ -173,6 +172,8 @@ handle_screenshot (XdpImplScreenshot *object,
 
   if (!g_variant_lookup (arg_options, "interactive", "b", &interactive))
     interactive = FALSE;
+  if (!g_variant_lookup (arg_options, "skip_dialog", "b", &skip_dialog))
+    skip_dialog = FALSE;
 
   if (arg_parent_window)
     {
@@ -193,7 +194,7 @@ handle_screenshot (XdpImplScreenshot *object,
   g_object_ref_sink (fake_parent);
 
   dialog = GTK_WINDOW (screenshot_dialog_new (arg_app_id,
-                                              screenshot_portal_version,
+                                              skip_dialog,
                                               interactive,
                                               shell));
   gtk_window_set_transient_for (dialog, GTK_WINDOW (fake_parent));
@@ -299,7 +300,6 @@ gboolean
 screenshot_init (GDBusConnection *bus,
                  GError **error)
 {
-  g_autoptr(GVariant) version_variant = NULL;
   g_autoptr(GVariant) result = NULL;
   GDBusInterfaceSkeleton *helper;
 
@@ -336,16 +336,6 @@ screenshot_init (GDBusConnection *bus,
                                         NULL,
                                         error);
 
-  version_variant = g_variant_get_child_value (result, 0);
-
-  if (!version_variant)
-    return FALSE;
-
-  g_variant_get_child (version_variant, 0, "u", &screenshot_portal_version);
-
-  if (screenshot_portal_version == 0)
-    return FALSE;
-
   g_debug ("providing %s", g_dbus_interface_skeleton_get_info (helper)->name);
 
   return TRUE;
diff --git a/src/screenshotdialog.c b/src/screenshotdialog.c
index b42251c..f2c3f5c 100644
--- a/src/screenshotdialog.c
+++ b/src/screenshotdialog.c
@@ -416,7 +416,7 @@ screenshot_dialog_class_init (ScreenshotDialogClass *class)
 
 ScreenshotDialog *
 screenshot_dialog_new (const char *app_id,
-                       uint32_t screenshot_portal_version,
+                       gboolean skip_dialog,
                        gboolean interactive,
                        OrgGnomeShellScreenshot *shell)
 {
@@ -441,7 +441,7 @@ screenshot_dialog_new (const char *app_id,
   gtk_label_set_label (GTK_LABEL (dialog->heading), heading);
 
   dialog->shell = g_object_ref (shell);
-  dialog->skip_dialog = screenshot_portal_version >= 3 && !interactive;
+  dialog->skip_dialog = skip_dialog;
 
   if (interactive)
     show_options (dialog);
diff --git a/src/screenshotdialog.h b/src/screenshotdialog.h
index 98b336c..eeea773 100644
--- a/src/screenshotdialog.h
+++ b/src/screenshotdialog.h
@@ -8,6 +8,6 @@ typedef struct _ScreenshotDialog ScreenshotDialog;
 typedef struct _ScreenshotDialogClass ScreenshotDialogClass;
 
 ScreenshotDialog * screenshot_dialog_new (const char *app_id,
-                                          uint32_t screenshot_portal_version,
+                                          gboolean skip_dialog,
                                           gboolean interactive,
                                           OrgGnomeShellScreenshot *shell);
```

on top of the above diffs. The point here is to not tie a xdg-desktop-portal version to a xdg-desktop-portal-gnome/xdg-desktop-portal-kde/... version.

Reply:
Maybe `skip_dialog` should be `permission_store_checked` or something instead? I assume that e.g. KDE did their own permission checking in the non-interactive path?

Reply:
That opens a couple of questions:

 * What should happen if the user denies the app the permission to take non-interactive screenshots? Should we continue to abort the screenshot call, or should it pass "skip_dialog=false"? (I think the former is the better option)
 * Should the version property should still be added?

Reply:
> I assume that e.g. KDE did their own permission checking in the non-interactive path?

AFAICS xdg-desktop-portal-kde is more liberal about that and doesn't show any dialogs in the non-interactive case anyway

Reply:
> * What should happen if the user denies the app the permission to take non-interactive screenshots? Should we continue to abort the screenshot call, or should it pass "skip_dialog=false"? (I think the former is the better option)

I'd say: if the backend has the new enough impl version, always check the permission store and never ask for a dialog, aka exactly what the new intended behavior is in this PR; if the backend is too old, assume it will take care of all permission checking, as is the case without this PR.

> 
> * Should the version property should still be added?

Only the backend/impl would need a version and version bump in the way I have in mind now.

The frontend/app facing API didn't change -> no API bump.

The backend/impl API did change -> API bump. Since there is no version, we need to add it to bump the API version.